src/components/KTableData/KTableData.vue

```diff
@@ -0,0 +1,591 @@
+        name="bulk-action-items"
+        :selected-rows="selectedRows"
+      />
+    </template>
+
+    <template
+      v-if="$slots['error-state']"
+      #error-state
+    >
+      <slot name="error-state" />
+    </template>
+
+    <template
+      v-if="$slots['empty-state']"
+      #empty-state
+    >
+      <slot name="empty-state" />
+    </template>
+
+    <template
+      v-if="$slots['empty-state-action-icon']"
+      #empty-state-action-icon
+    >
+      <slot name="empty-state-action-icon" />
+    </template>
+
+    <template
+      v-for="slot in getHeaderSlots"
+      :key="slot"
+      #[slot]="{ column }"
+    >
+      <slot
+        :column="column"
+        :name="slot"
+      />
+    </template>
+
+    <template
+      v-for="slot in getHeaderTooltipSlots"
+      :key="slot"
+      #[slot]="{ column }"
+    >
+      <slot
+        :column="column"
+        :name="slot"
+      />
+    </template>
+
+    <template
+      v-for="slot in getCellSlots"
+      :key="slot"
+      #[slot]="{ row, rowKey, rowValue }"
+    >
+      <slot
+        :name="slot"
+        :row="row"
+        :row-key="rowKey"
+        :row-value="rowValue"
+      />
+    </template>
+
+    <template
+      v-if="$slots['action-items']"
+      #action-items="{ row }"
+    >
+      <slot
+        name="action-items"
+        :row="row"
+      />
+    </template>
+
+    <template
+      v-if="$slots['row-expanded']"
+      #row-expanded="{ columnWidths, nestedHeaders, row }"
+    >
+      <slot
+        :column-widths="columnWidths"
+        name="row-expanded"
+        :nested-headers="nestedHeaders"
+        :row="row"
+      />
+    </template>
+  </KTableView>
+</template>
+
+<script setup lang="ts">
+import type { Ref } from 'vue'
+import { ref, watch, computed, onMounted, useSlots } from 'vue'
+import KTableView from '@/components/KTableView/KTableView.vue'
+import useUtilities from '@/composables/useUtilities'
+import type {
+  TablePreferences,
+  TableDataHeader,
+  TableColumnSlotName,
+  TableColumnTooltipSlotName,
+  SortColumnOrder,
+  TableSortPayload,
+  ButtonAppearance,
+  RowLink,
+  TablePaginationAttributes,
+  PageChangeData,
+  PageSizeChangeData,
+  TableState,
+  TableStatePayload,
+  SwrvState,
+  SwrvStateData,
+  SortHandlerFunctionParam,
+  TableProps,
+  TableDataFetcherParams,
+} from '@/types'
+import { EmptyStateIconVariants } from '@/types'
+import useUniqueId from '@/composables/useUniqueId'
+
+interface TableDataProps extends TableProps {
+  fetcher: (param: TableDataFetcherParams) => Promise<any>
+  /**
+   * A prop to pass in an array of headers for the table
+   */
+  headers: TableDataHeader[]
+  fetcherCacheKey?: string
+  cacheIdentifier?: string
+  searchInput?: string
+  initialFetcherParams?: TableDataFetcherParams
+  clientSort?: boolean
+  sortHandlerFunction?: (param: SortHandlerFunctionParam) => Record<string, any>[]
+  sortable?: boolean
+  hidePaginationWhenOptional?: boolean
+  hideToolbar?: boolean
+}
```

Comment:
nit: can we move this to the types file?

Reply:
Done